### PR TITLE
[chore] change some fn's and variable's name

### PR DESCRIPTION
**Pull request by GG2002**
Change some fn's and variable's name, and add `serde` and `graphviz` features. They're small changes.

### CODE REVIEW

File: src/entry.rs [hunk 1]
```diff
@@ -5,7 +5,10 @@ use crate::node::Node;
 
 /// A view into a single entry in a map, which may either be vacant or occupied.
 #[derive(Debug)]
-pub enum Entry<'a, T, V, Ix> {
+pub enum Entry<'a, T, V, Ix>
+where
+    T: Ord,
```
Comment (bsbds) [inline src/entry.rs]: It's redundant to add trait bounds here.

File: src/entry.rs [hunk 2]
```diff
@@ -15,17 +18,23 @@ pub enum Entry<'a, T, V, Ix> {
 /// A view into an occupied entry in a `IntervalMap`.
 /// It is part of the [`Entry`] enum.
 #[derive(Debug)]
-pub struct OccupiedEntry<'a, T, V, Ix> {
+pub struct OccupiedEntry<'a, T, V, Ix>
+where
+    T: Ord,
```
Comment (bsbds) [inline src/entry.rs]: Same as above.

File: src/entry.rs [hunk 3]
```diff
@@ -15,17 +18,23 @@ pub enum Entry<'a, T, V, Ix> {
...
+    pub node_idx: NodeIndex<Ix>,
 }
 
 /// A view into a vacant entry in a `IntervalMap`.
 /// It is part of the [`Entry`] enum.
 #[derive(Debug)]
-pub struct VacantEntry<'a, T, V, Ix> {
+pub struct VacantEntry<'a, T, V, Ix>
+where
+    T: Ord,
```
Comment (bsbds) [inline src/entry.rs]: Same as above.

File: src/intervalmap.rs [hunk 4]
```diff
@@ -2,11 +2,27 @@ use crate::entry::{Entry, OccupiedEntry, VacantEntry};
...
+#[cfg(feature = "graphviz")]
+use std::io::Write;
 
 /// An interval-value map, which support operations on dynamic sets of intervals.
+#[cfg_attr(feature = "serde", derive(Serialize, Deserialize))]
 #[derive(Debug)]
-pub struct IntervalMap<T, V, Ix = DefaultIx> {
+pub struct IntervalMap<T, V, Ix = DefaultIx>
+where
+    T: Ord,
```
Comment (bsbds) [inline src/intervalmap.rs]: Same as above.

File: src/intervalmap.rs [hunk 5]
```diff
@@ -242,11 +251,7 @@ where
     #[inline]
     #[must_use]
     pub fn new() -> Self {
-        Self {
-            nodes: vec![Self::new_sentinel()],
-            root: Self::sentinel(),
-            len: 0,
-        }
+        Self::with_capacity(0)
```
Comment (bsbds) [inline src/intervalmap.rs:252]: Why is the sentinel node removed?

File: src/intervalmap.rs [hunk 6]
```diff
@@ -719,423 +663,149 @@ where
...
-        stack.extend(Self::left_link(
-            self.map_ref,
-            self.map_ref.node_ref(x, Node::right),
-        ));
-        Some(self.map_ref.node_ref(x, |xn| (xn.interval(), xn.value())))
-    }
-}
-
-#[cfg(test)]
-mod test {
```
Comment (bsbds) [inline src/intervalmap.rs:877]: Why removes the tests?

File: src/entry.rs (hunk 1 above)
Comment (GG2002) [reply src/entry.rs]: I thought `Entry` and other struct below are used by `pub use`, so perhaps adding a bound is safer?

File: src/intervalmap.rs (hunk 6 above)
Comment (GG2002) [reply src/intervalmap.rs:877]: These tests will be presented at next pr.

File: src/intervalmap.rs (hunk 5 above)
Comment (GG2002) [reply src/intervalmap.rs:252]: I thought we could replace it with `Node::new_sentinel` and `Node::new`, what do you think about it?

File: src/entry.rs (hunk 1 above)
Comment (bsbds) [reply src/entry.rs]: Adding a bound to the struct itself requires repeating it in every implementation. Alternatively, you can add a trait bound to a specific implementation, which restricts users from utilizing that implementation unless they satisfy the bound.

Please also refer to: https://stackoverflow.com/questions/49229332/should-trait-bounds-be-duplicated-in-struct-and-impl

File: src/index.rs [hunk 7]
```diff
@@ -1,30 +1,49 @@
...
-pub unsafe trait IndexType: Copy + Default + Hash + Ord + fmt::Debug + 'static {
+pub trait IndexType: Copy + Default + Hash + Ord + fmt::Debug + 'static {
+    const SENTINEL: Self;
     fn new(x: usize) -> Self;
     fn index(&self) -> usize;
     fn max() -> Self;
+    fn is_sentinel(&self) -> bool {
+        *self == Self::SENTINEL
+    }
 }
```
Comment (bsbds) [inline src/index.rs]: Please add docs to the index type trait and it's methods.

File: src/intervalmap.rs (hunk 5 above)
Comment (bsbds) [reply src/intervalmap.rs:252]: Sorry, I didn't see it moved to the `with_capacity` method.

File: src/entry.rs (hunk 1 above)
Comment (GG2002) [reply src/entry.rs]: I agree with you.

Comment (GG2002) [reply src/entry.rs]: Unfortunately, I found these bounds were needed by serde feature, more precisely, it's needed by
```rust
#[cfg(feature = "serde")]
impl<'de, T: Deserialize<'de> + Ord> Deserialize<'de> for Interval<T> {
    fn deserialize<D: Deserializer<'de>>(deserializer: D) -> Result<Self, D::Error> {
        let (low, high) = <(T, T)>::deserialize(deserializer)?;
        Ok(Interval::new(low, high))
    }
}
```
I thought after deserialization, T of Interval<T> should meet Ord at least. And this leads to a lot of explicit bounds. What do you think about it?

**Review by bsbds — APPROVED** [review]

**Pull request merged**